### PR TITLE
threadtree: refactor logging's Context into ThreadContext

### PR DESCRIPTION
**Pull request by koreno**
feel free to suggest alternate names there

### CODE REVIEW

File: easypy/threadtree.py [hunk 1]
```diff
@@ -258,3 +261,116 @@ def dump_threads():
...
+        with TC(i=1):
+            assert TC.i == 1
+            assert TC.j == 0
+
+            with TC(i=1, j=1):
+                assert TC.i == 2
+                assert TC.j == 1
+
+
+    'concatenated': attributes named here get concatenated into a list, instead of overwritten:
```
Comment (tigrawap) [inline easypy/threadtree.py]: concatenated usually used for strings...  I'd say "merged", "joined", "appended"

File: easypy/threadtree.py [hunk 2]
```diff
@@ -258,3 +261,116 @@ def dump_threads():
...
+        # (thread-b)
+        assert TC.my_data == 'b'
+        with TC(my_data='c'):
+            spawn_thread_c()
+        assert TC.my_data == 'b'
+
+        # (thread-c)
+        assert TC.my_data == 'c'
+
+    'accumulated': attributes named here get incremented each time, instead of overwritten:
```
Comment (tigrawap) [inline easypy/threadtree.py]: 1) accumulated - incremental?
2) MB add `depth` as common incremental attribute? Sounds like often need for cases when such context would be used

Comment (koreno) [inline easypy/threadtree.py]: 1. I could change them to `append` and `increment`
2. hmmm, we haven't needed that ourselves, so it doesn't seem like a common need. different apps might want to increase depth differently too

File: easypy/concurrency.py [hunk 3]
```diff
@@ -174,7 +174,11 @@ def _get_buffer(self, **kw):
                     buff.write(exc.render(**kw))
                 else:
                     if hasattr(exc, "context"):
-                        context = "(%s)" % ", ".join("%s=%s" % p for p in sorted(exc.context.items()))
+                        context = exc.context.copy()
+                        context.pop("indentation", None)
+                        if 'context' in context:  # 'context' should be renamed 'breadcrumbs'
+                            context['context'] = "%s" % ";".join(context['context'])
```
Comment (idanarye) [inline easypy/concurrency.py]: The `"%s" % ` is redundant

**Pull request merged**